controlplane/kubeadm/internal/controllers/controller.go

```diff
@@ -791,6 +810,88 @@ func (r *KubeadmControlPlaneReconciler) reconcileEtcdMembers(ctx context.Context
 	return nil
 }
 
+func (r *KubeadmControlPlaneReconciler) reconcilePreTerminateHook(ctx context.Context, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+	if !controlPlane.HasDeletingMachine() {
+		return ctrl.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// Pick the Machine with the oldest deletionTimestamp to keep this function deterministic / reentrant
+	// so we only remove the pre-terminate hook from one Machine at a time.
+	deletingMachine := controlPlane.DeletingMachines().OldestDeletionTimestamp()
+	log = log.WithValues("Machine", klog.KObj(deletingMachine))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Return early if there are other pre-terminate hooks for the Machine.
+	// The KCP pre-terminate hook should be the one executed last, so that kubelet
+	// is still working while other pre-terminate hooks are run.
+	if machineHasOtherPreTerminateHooks(deletingMachine) {
```

Comment:
I think we should somehow document that KCP controlled machines are adding a hook that expects to be always run as last (to ensure kubelet is running). 

This should go in the PR description first but probably in a few other places too (the latter could be part of this PR or a quick follow-up, up to you); Probably this should go:
- At the end of the go comment for PreTerminateDeleteHookAnnotation
- 1.9 upgrade instruction
- release notes for the patch release 

(and we should bring this up at the office hours)

Reply:
Sounds like a follow-up to me

Reply:
Maybe you can take that one. Probably easier because you know better where you want to add it :)

Reply:
Opened a follow-up PR (https://github.com/kubernetes-sigs/cluster-api/pull/11153) + added it to the agenda